### DISCUSSION TITLE
[Question] When is docfx v3 scheduled for release?

### DISCUSSION THREAD

**Discussion by zorngiebel** [dotnet/docfx#8277] [General]
Hi, 

are there any news regarding a planned release date for docfx v3? 

Right now we are thinking about using v3 for our documentation, but if it is still far away we might simply go ahead and use v2.

Especially important for us is generating html and pdf files from Markdown.

Thanks a lot!

**Comment by FSharpCSharp** [comment]
That is indeed a very valid question. I'm not even sure if the roadmap is still up to date. It was updated more than half a year ago, in July last year. It would be nice to know if the release is in the distant or near future.

**Comment by zorngiebel** [comment]
Are there any news? As FSharpCSharp stated, the roadmap really seems outdated, so we are not sure about the state of things and even if there is progress at all.

**Comment by thuannguy** [comment]
I have asked this questions a couple of times. Any news?

**Comment by abarger-bss** [comment]
We are also interested in an update so we can plan to allocate time for the migration.

**Comment by stap123** [comment]
If I had to guess, given the number of beta releases being triggered at the moment, I'd say they're going to drop it at BUILD so they're staying quiet so they don't ruin the announcement. Would be nice to get a response though 👀

**Comment by mansellan** [comment]
Would love to get involved with DocFx, but can't justify the effort until v3 drops.

**Comment by yufeih** [comment]
We've just had other priorities that have taken precedence over a formal v3 release. We still have a bit of work to do on it before it's publicly consumable, but we expect to have it done later in the summer.

The [roadmap](https://github.com/dotnet/docfx/blob/v3/docs/roadmap.md) now describes the present state and the gaps of v3.

**Comment by idg10** [comment]
Is any of the outstanding work something that could be helped forward with community contributions? Or is it all stuff that basically needs to be done by team members?

**Comment by Sotty75** [comment]
Hi guys,
did you manage to start working on this? I need to decide which platform to move to (i am on docusaurus now) and DocFX v3 has precedence, not sure if i can wait for much longer though.

**Comment by dovholuknf** [comment]
I'm in the same boat... I've started the process of moving from docfx to "something else". v3 really needs to drop...

**Comment by Sotty75** [comment]
> I'm in the same boat... I've started the process of moving from docfx to "something else". v3 really needs to drop...

I would just be happy to get a status update more often, otherwise people looks at this as a dead project.

**Comment by Joren-Thijs** [comment]
Any updates? Can we expect v3 to drop with .NET 6 in November or after?
I just started looking into v2 and although it works I would like to know if this project has a real future?

Second I also agree with @Sotty75.
I understand if it can't be finished in a snap and teams are busy with all the net6/WinUI3/MAUI en VS2022 stuff going on but it would be great to be kept up to date.

Third docfx is a really awesome and powerfull tool. And since almost all developers (including me) hate writing documentation, Microsoft has the opportunity to offer a great tool here with v3. So i think it's a bit of a shame it doesn't have a higher priority.

**Comment by JeremyCaney** [comment]
I'm eager for an update on this as well. I'm anxious to adopt DocFx for a couple of projects, but with lack of support for [key C# 9 features](https://github.com/dotnet/docfx/issues/7191) (such as records) and [new C# 10 features](https://github.com/dotnet/docfx/issues/7737) (such as global using directives), it's not currently possible without sourcing from the assemblies. 

Alternatively, are we reaching a point where the beta releases are stable enough to deploy to NuGet? At minimum, it'd be nice to begin getting started with the latest release, and potentially report bugs as appropriate, without needing to e.g. check an independently maintained copy of the binaries into source code.

Thank you for your hard work.

**Comment by red8888** [comment]
Is v3 going to support .net core 5? https://github.com/dotnet/docfx/issues/7022

**Comment by KalleOlaviNiemitalo** [comment]
DocFX v3 currently targets .NET 6.0, so I don't think you'll be able to run it on .NET 5.
<https://github.com/dotnet/docfx/blob/ba6dcef726b9b8dd610373925b5198ada7bd656f/Directory.Build.props#L17>

It should still be usable for generating documentation from .NET 5 assemblies, though.

**Comment by actionthomas** [comment]
> Is v3 going to support .net core 5? #7022

As .net 5 EOL is May 8, 2022 , I don't think they'll even bother.

**Comment by dovholuknf** [comment]
> > Is v3 going to support .net core 5? #7022
> 
> As .net 5 EOL is May 8, 2022 , I don't think they'll even bother.

I don't think they'll ever bother to release 3.0 either.... oh you were referring to .net core 5 support... silly me... #waiting-forever

I really do want to like docfx but the glacial pace has made me question using it :(

**Comment by Joren-Thijs** [comment]
Yeah, the only thing that keeps me hopefull is that PR's keep being merged for V3 instead of the repo going dead like some other projects

**Comment by Sotty75** [comment]
I made myself the idea this is such a low priority project that they have not a schedule for the roadmap, but still willing to release at some point in time. Unfortunately, this situation is frustrating as they did set some expectations wich are being missed.

**Comment by mansellan** [comment]
Hopeful that this project is not dead, but a little more communication would be most welcome.

**Comment by FSharpCSharp** [comment]
The project is obviously being worked on at the moment. I would be personally interested to know what the current status is. The roadmap itself was updated almost half a year ago, and I unfortunately have no overview of what is still missing or how far away we are before the real release. Feedback would indeed be quite nice.

Maybe @yufeih can say something about it, or even update the roadmap, as he seems to be doing the main work in this project.

**Comment by yufeih** [comment]
We may have something to share next quarter. Meanwhile I have an early prototype [here](https://docascode.github.io/template/api/catlibrary/). To set expectations, this is just an early prototype.

**Comment by Sotty75** [comment]
Thanks a lot, the prototype looks already pretty good.
I am glad we got some updates :)

**Comment by thuannguy** [comment]
Thank you for an update (at last) 😄 
To be honest, I'm not optimistic. "Something to share next quarter" does sound like they will share about their updated roadmap/timeframe etc. which might mean that a 3.0 release would be many quarters away.

**Comment by idg10** [comment]
As far as I can tell, this bit in the roadmap:

> API Reference: Microsoft Docs uses a separate set of tools to ingest API references for a variety of programming languages in addition to C#.

is really saying "DocFX no longer generates API reference documentation."

I understand the argument that by separating this part out, it makes DocFX a more general-purpose tool. But the problem is, this API reference documentation was one of the main reasons for using DocFX in the first place!

Is this "separate set of tools" is publicly available? I've not been able to find it? And if not, what are we expected to do if we want to produce API Reference docs?

**Comment by dpeter99** [comment]
Well the wait continues..... #waiting-forever

I as well have been putting off using DocFx (and using alternatives) for years now, as V3 slowly comes around.

Am I seeing it correctly that currently, it can not actually generate code documentation and only the .md generation is in place?
If this is not the case is there a publicly available sample to show how to do it?

Is there a timeline currently?

**Comment by lanthonyneville** [comment]
@idg10 - if "DocFX no longer generates API reference documentation.", then how come @yufeih's early prototype [here] shows API documentation? (https://docascode.github.io/template/api/catlibrary/).

**Comment by dpeter99** [comment]
As far as I understand they have code that generates the metadata files that are then consumed by docfx.
This would mean that docfx is a more traditional static site generator and that multiple language support becomes way easier.
These codebases however don't seem to be public yet.

**Comment by FSharpCSharp** [comment]
> We may have something to share next quarter. Meanwhile I have an early prototype [here](https://docascode.github.io/template/api/catlibrary/). To set expectations, this is just an early prototype.

I don't want to seem rude in any way, but am I missing something? Shouldn't there possibly be news about the release in the current quarter? We're already in the second third of this one, and so far I haven't read anything more about it. Has the release been postponed again, or is there another place where this announcement was made?

**Comment by KalleOlaviNiemitalo** [comment]
[README.md](https://github.com/dotnet/docfx/blob/97c4493c83b1a71e455c60265de56f28b7aee686/README.md#license) says "DocFX is supported by the .NET Foundation." However, it's not listed in [.NET Foundation Projects](https://dotnetfoundation.org/projects). Was it already retired as not satisfying the [Eligibility or Health requirements](https://dotnetfoundation.org/projects/net-foundation-project-committee#eligibility-criteria)?

**Comment by FSharpCSharp** [comment]
@yufeih 

I don't want to seem rude or pushy in any way, but time is running out, so are the quarters, and since your announcement in February with the reference to the coming quarter, unfortunately nothing more has been heard about version 3. Prototypes appear regularly now, but you can't see where the current state of affairs is. PDF still doesn't seem to go in these, a doc on how to use it doesn't seem to exist either. So slowly but surely I wonder if I will even live to see the release.

That's why I ask you for a statement about the state of affairs, how things are going, also connected with the request to bring the roadmap up to date.

**Comment by Sotty75** [comment]
> We may have something to share next quarter. Meanwhile I have an early prototype [here](https://docascode.github.io/template/api/catlibrary/). To set expectations, this is just an early prototype.

Is there any update?

**Comment by bitbonk** [comment]
@yufeih Looks like "next quarter" didn't work out. Dou have any updates for us on how your plans look like?

**Comment by yufeih** [comment]
Unfortunately, the plan didn't work out. @herohua or @markjulmar may have more info.

**Comment by josesimoes** [comment]
Adding my request here for an official update on how things are going regarding the release of v3. 
No need for a blog post!! Just a couple of sentences explaining the current state of affairs and what can we expect regarding the release. Plain. Simple. 
And thanks in advance! 😄

**Comment by bitbonk** [comment]
> Adding my request here for an official update on how things are going regarding the release of v3. 
> 
> No need for a blog post!! Just a couple of sentences explaining the current state of affairs and what can we expect regarding the release. Plain. Simple. 

In case the current state of affairs has changed in any way. *wink *wink

**Comment by lanthonyneville** [comment]
Can anyone say why this project is now read only?

**Comment by lukaszch** [comment]
Tired of waiting for any progress, we switched to Docusaurus, and we are super happy with the decision.

**Comment by bitbonk** [comment]
I think it should be apparent to anyone following the issues in this repo, that docfx is no longer meant for public consumption and that it solely exists to support <https://learn.microsoft.com>. The maintainers have basically stopped all communication with the community for several years now.

Here are a some potential alternatives:
- https://docusaurus.io
- https://www.statiq.dev/docs or its predecessor https://wyam.io
- https://asciidoc.org

**Comment by bitbonk** [comment]
@lukaszch 
> Tired of waiting for any progress, we switched to Docusaurus, and we are super happy with the decision.

Have you figured out a good way to generate and include API docs into Docusaurus?

**Comment by lukaszch** [comment]
@bitbonk,
We didn't need to generate API docs yet.
However, if needed, we will probably convert Visual Studio XML documentation files to markdown with a tool like this: https://github.com/lijunle/Vsxmd, and include those markdown files in Docusaurus.

**Comment by dovholuknf** [comment]
I did exactly the same. docfx is dead to me which bummed me out. It does some stuff better than docusaurus but docusaurus has been better overall for numerous other reasons (notably the live reload). 

I ended up using docfx to generate the static content for c# in my c# project. If you want you can then add those html files to the docusaurus site... That's basically what I did (not quite, but close enough). I didn't restyle that doc. it's ugly and very visually different but for now, it works.

**Comment by markjulmar** [comment]
Sorry for the late response; I missed the tag earlier, thanks @yufeih! First apologies for the long response, we've been dark for a while, and we finally have a way forward to talk about.

DocFX is _not_ dead! It's been successfully used to create all sorts of internal and external static websites, including the original incarnation of Microsoft docs. In addition, there is still a community dedicated to supporting and improving this project - I know we've been quiet lately; life events and other changes have sidetracked several of us, but we're still around!

Regarding the V3 branch -- that branch has been around for a looong time, and it was the primary place where a bunch of experimentation and work was happening - most of it to support <learn.microsoft.com>, as someone noted earlier. Officially, we (Learn PM and Engineering) aspired to make this an official release. However, as time has passed and the Learn site has evolved into something entirely different, we've realized it didn't make sense to continue down this road. Instead, our community efforts will focus on investing in the main branch and pulling some of the interesting ideas into it from V3 - more on that in a sec.

To provide a little background: as you all know, DocFX was an instrumental tool in allowing Microsoft to build up a massive documentation site that supported various content types/schemas and a full build tool capable of generating an entire HTML/JS/CSS site from basic config and Markdown. Microsoft could not have created Docs without it. However, as Docs (now Learn) has evolved and added more dynamic content and capabilities, we found it more challenging to continue to rely on static page generation. A significant amount of code ended up in the client-side JavaScript to enable dynamic features, which added complexity, code conflicts, and in some cases, reduced page performance due to all the round trips to the server and DOM manipulation.

The site has been moving to a dynamic server-side composition model to address these problems. This composition layer allows the pages to be dynamically constructed based on the environment, signed-in identity, and other runtime state information. This approach reduces the client-side JS significantly and has allowed us to innovate in several areas of the site. As you can imagine, this also has the side effect of dramatically reducing Learn's dependency on DocFX to generate content. It's still present, but the build system only uses pieces of it, and the produced output is more of an intermediate format than the final HTML.

Given these changes in the publishing model, it made sense to stop the official development of V3 - so you won't see any more work on that branch for the Learn site itself. **BUT** - DocFX will continue to evolve! Several of the key contributors are discussing what's next. As mentioned above, we'll cherry-pick features and structural changes from the V3 branch (that version breaks several key scenarios we want to keep). The official roadmap (once it's decided) will determine how much gets pulled into the main branch, so stay tuned for that. If there's something in V3 that you found particularly exciting or think would be useful, let us know.

Finally, building off that last statement, we are looking for additional contributors to the project to help us plan and implement new features and capabilities. If anyone is interested, please engage with the team in the discussion forums once the roadmap is up!

**Comment by markjulmar** [comment]
I'm going to close this issue - but it doesn't mean it has to be the end of the conversation!

**Comment by darraghjones** [comment]
Where does this leave the docfx dotnet tool and .net core support?

**Reply by yufeih** [reply]
We are looking at various approaches to move to .NET core. One of the most challenging areas is to figure out a way forward with the managed reference pipeline: it is today using roslyn to reflect source code and assemblies, which presents many challenges in keeping up with csharp/vs update as well as moving to .NET core.

**Comment by Panda-Sharp** [comment]
> I'm going to close this issue - but it doesn't mean it has to be the end of the conversation!

@markjulmar thanks for your reply! maybe could be worth to move this issue to discussions?
